### PR TITLE
Add: Buttons block

### LINKED ISSUE #16480: Button Block and Multiple Buttons
One fairly common need is aligning more than one button in a row. Multiple block libraries have added similar tools (either a special multiple buttons block or the ability to add more instances in a single button block). We should explore bring this into core, likely leveraging child blocks for the extra buttons, but in a more convenient way than forcing people to add columns first to hold each button in place.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
Closes: https://github.com/WordPress/gutenberg/issues/16480
This PR adds a buttons block. A container for buttons. It is starting point right now it is not much more than a container but we can follow up from this and discussi new features to add.

## How has this been tested?
I tested the buttons and button block work as expected by doing some smoke tests.

## Screenshots 
Sep-05-2019 22-23-55

### CODE REVIEW

**Comment by paaljoachim** [comment]
It seems really weird to have a button and a buttons block. It would be better to improve the existing button block and rename it to buttons.

**Comment by mtias** [comment]
`Button` is expected to become a child block of `Buttons`, so the inserter would only expose `Buttons` by default. Just need to ensure nothing weird happens if you use `Button` directly in a template or existing content.

**Comment by jorgefilipecosta** [comment]
> Button is expected to become a child block of Buttons, so the inserter would only expose Buttons by default. Just need to ensure nothing weird happens if you use Button directly in a template or existing content.


Hi @mtias, this PR was updated and now follows this logic. Only buttons block is available, button can still be used in existing content and existing templates.

**Comment by mtias** [comment]
> Is this being backported to WP 5.3?

No, for next release. Too many changes already.

File: packages/block-library/src/buttons/editor.scss
Comment (mtias) [inline packages/block-library/src/buttons/editor.scss]: Can we extract all of this to `InnerBlocks horizontalDisplay`?

Comment (jorgefilipecosta) [reply packages/block-library/src/buttons/editor.scss]: I extracted this CSS logic into generic InnerBlocks flags.

**Comment by jorgefilipecosta** [comment]
I performed some updates on this PR, following some suggestions by @mtias. I moved the movers into the block toolbar as an experiment, and now the standard block UI is hidden using generic inner blocks flags added in https://github.com/WordPress/gutenberg/pull/18173.

File: packages/block-library/src/editor.scss
Comment (jasmussen) [inline packages/block-library/src/editor.scss]: Is this necessary? I suspect this will cause a lot of editor styles to break because it adds needless extra specificity to editor styles.

File: packages/edit-post/src/components/visual-editor/style.scss
Comment (jasmussen) [inline packages/edit-post/src/components/visual-editor/style.scss]: Not sure what this does, but it means all full-wide blocks now have margin left and right of them:

Screenshot 2019-11-01 at 15 58 34

File: packages/block-editor/src/components/block-list/style.scss
Comment (jasmussen) [inline packages/block-editor/src/components/block-list/style.scss]: I definitely emphathise with the complex block paddings. But this PR seems to go a bit farther than necessary in making things work. Perhaps it's worth waiting for #17877 to land before rebasing this on top? 

Alternately, and I suspect this is due to the onerous extra UI applied to child blocks, you could snatch the code I wrote for social links, that should allow you to change the margins and paddings just for this one block, while awaiting better improvements than adding a new `has-spacing` property. Specifically it starts here: https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/social-links/editor.scss#L14

Try the social links block and see how child blocks are well spaced out. If you need my help feel free to ping and I'll take a look.

**Comment by jorgefilipecosta** [comment]
Hi @jasmussen thank you for reviewing this PR. The specific changes you mentioned are part of a PR this PR depends on https://github.com/WordPress/gutenberg/pull/18173. I moved the discussion to that PR.

**Comment by jorgefilipecosta** [comment]
This PR and https://github.com/WordPress/gutenberg/pull/18173 were both rebased.

File: packages/block-library/src/button/edit.js
Comment (youknowriad) [inline packages/block-library/src/button/edit.js]: Instead of these custom movers, I think we should just use the vertical movers used by the navigation block.

File: packages/block-library/src/button/edit.js
Comment (youknowriad) [inline packages/block-library/src/button/edit.js]: This also could use the new LinkControl introduced by the Navigation block. This could be a separate PR though.

File: packages/block-library/src/buttons/edit.js
Comment (youknowriad) [inline packages/block-library/src/buttons/edit.js]: A lot of these are unnecessary now right?

File: packages/block-library/src/buttons/edit.js
Comment (youknowriad) [inline packages/block-library/src/buttons/edit.js]: If I understand correctly this is used to keep the inline URL component for the Button block when used like today. Is it really necessary, can we just always use the inline UI, It seems way better to me.

File: packages/block-library/src/buttons/edit.js
Comment (youknowriad) [inline packages/block-library/src/buttons/edit.js]: @draganescu worked on a PR where if there's only one allowed child block, the appender is automatically something very similar to the one used here. Can't we just rely on this behavior?

Also, I feel this is also very similar to the Navigation block, how is this solved there?

**Review by youknowriad — COMMENTED** [review]
Left some comments, but great job so far on this PR.

File: packages/block-library/src/buttons/index.js
Comment (Soean) [inline packages/block-library/src/buttons/index.js]:
```suggestion
	description: __( 'Prompt visitors to take action with a group of button-style links.' ),
```

File: packages/block-library/src/button/edit.js
Comment (jorgefilipecosta) [reply packages/block-library/src/button/edit.js]: Right, this was added as an experiment to compare how movers in the toolbar would look like. Now inner blocks already supports movers in the toolbar so I removed this code and relied on InnerBlocks.

File: packages/block-library/src/button/edit.js
Comment (jorgefilipecosta) [reply packages/block-library/src/button/edit.js]: I updated the PR to use LinkControl, so we have more similar UI's.

File: packages/block-library/src/buttons/edit.js
Comment (jorgefilipecosta) [reply packages/block-library/src/buttons/edit.js]: Yes, this code is not necessary as this behavior is now the default one, if only one block is allowed it is inserted right away. This code was added before this functionality was merged and it was now removed.

File: packages/block-library/src/buttons/edit.js
Comment (youknowriad) [inline packages/block-library/src/buttons/edit.js]: Nit: this could be just `="horizontal"`

File: packages/block-library/src/button/edit.js
Comment (youknowriad) [inline packages/block-library/src/button/edit.js]: The  API is a bit weird to me. That's feedback that is not meant directly to this  PR but more about the LinkControl component. cc @getdave 
I think ideally,  the current link prop is just another setting and we should have a single onChange handler. Any particulra reason to separate both?
I also feel we should separate the definition of the settings (id, title) from the value (checked). 
Should we also absorb `handleLinkControlOnKeyDown` inside the component itself?

**Review by youknowriad — APPROVED** [review]
Nice work here 👍

File: packages/block-library/src/button/edit.js
Comment (youknowriad) [reply packages/block-library/src/button/edit.js]: I took a stab at this here https://github.com/WordPress/gutenberg/pull/19396

**Comment by aduth** [comment]
I've been seeing quite a few intermittent build failures recently which appear to stem from the test file introduced in this pull request.

Example: https://travis-ci.com/WordPress/gutenberg/jobs/272683577

**Comment by ellatrix** [comment]
The alignment buttons don't work. See #19616.

File: packages/block-library/src/button/edit.js
Comment (aduth) [inline packages/block-library/src/button/edit.js]: What purpose does this serve? How would a future maintainer be expected to interpret this requirement to avoid a regression?

A few suggestions:

- `Event#stopPropagation` is a code smell, and we should avoid it as much as possible
- When we have no alternative, we should document exhaustively why we're stopping propagation. Something like an inline comment or, in this case, even just naming the function in such a way which communicates the intention.

**Comment by cinghaman** [comment]
@aduth do you know if this is going to be added to the core gutenberg anytime soon?

**Comment by aduth** [comment]
@cinghaman This is already available in the latest version of the plugin.

See: https://make.wordpress.org/core/2020/01/09/whats-new-in-gutenberg-8-january/

Based on the regular development schedule, it would be expected this should be available in the next major release of WordPress (5.4).

**Pull request merged**
